### ISSUE TITLE
USB external drive not showing correctly and files not accessible

### ISSUE THREAD

**Issue by Neverends4** [nova-video-player/aos-AVP#301]
Environment: Toshiba (OEMed by TCL) **Android 6.0 TV** with Mstar chips
Version tested: **v4.35 release and v5.7 SMBv2 beta release**
Detail: Attached a FAT32 USB drive to TV and Nova can neither display it in main menu nor access it.
The only way to get it displayed is through:
**"Legancy interface"-"Cloud"** and choose system file manager, but this only "display" the external drive, when openning any media files on it Nova reports error:
![ex usb-1.jpg](https://i.loli.net/2020/02/21/WPdeNUA96XIZ3ir.jpg)

But on the same TV the drive can be accessed through built-in file manager, many 3rd party file explorers (such as Solid explorer, Mixplorer), and the media can be opened and played normally by Nova if triggered within file explorer apps:
1. File address shown by Mixplorer:
![mix-file url.jpg](https://i.loli.net/2020/02/21/jbsVa1nlK5LtveH.jpg)

2. Status when opened within Mixplorer using Nova (can be played normally):
![open with fe.jpg](https://i.loli.net/2020/02/21/jxU97HPOt2bNWV4.jpg)

Last but not least, the same USB drive can be displayed normally using Archos, tested with newest version 10.2-20180416:

Main interface in Archos:
![archos.jpg](https://i.loli.net/2020/02/21/VHMQpsc4RZFqPA1.jpg)

And file address reported by Archos:
![archos2.jpg](https://i.loli.net/2020/02/21/1yaOlDoTA2HEs4f.jpg)

**Comment by courville** [comment]
This is strange since when a USB device is plugged nova should receive  UsbManager.ACTION_USB_DEVICE_ATTACHED.
I have reworked the original Archos code to be compliant with earlier Android releases (no choice) since it was using forbidden reflections.
Could you please try out this apk that has more logs and report an `adb logcat` (ideally when nova is launched `adb logcat --pid=$(adb shell pidof -s org.courville.nova`) and attach your USB drive to see what happens?
https://drive.google.com/open?id=15Rj7lh8M891N7KbakfRgmkJp_C7K03ja
Note that I tried with the nvidia shield and it works as intended.
Thanks.

**Comment by Neverends4** [comment]
Hi Courville, used the app you gave but seems there's something wrong with my TV's logcat (note I do not have root access and in fact I tried and cannot get it rooted), tried
`adb logcat --pid=$(adb shell pidof -s org.courville.nova) ` and `adb shell logcat org.courville.nova > d:\log.txt` and neither shows any result. I tried these two commands on my rooted phone (pie, 3rd party rom), both commands runs correctly.
If you could add bugreport within Nova then maybe I can try another time :)
Sorry I could not provide the log. Thing I can confirm: This bug is also reported by other users using the same TV model as me.

**Comment by courville** [comment]
You need to connect to tv via wifi (adb connect IPADDR) first after enabling adb in the TV settings in developers settings (no need to root the tv).
Then an adb -s IP:5555 logcat will do.

**Comment by Neverends4** [comment]
Yeah, I totally understand what you mean, and the above was done already connected through wifi and adb enabled. I even tried installed some logcat apps on TV such as Logcat Reader and Logcat no-root and gave them READ_LOG permission via adb, still the same, not even A SINGLE LINE of log, not at all... restared TV several times, the same. I believe the manufacturer deliberately disable logcat output in the ROM.

**Comment by Neverends4** [comment]
I used an OTG adapter with one USB drive (FAT32, 16GB) on my rooted phone to get log captured. [Link to log here, expire on Mar. 3](https://we.tl/t-LXlWxw2iYy)
On the phone Nova still cannot read the drive, even when it did show an "External Drive" icon. When entered  Nova displayed "File not found" even when I chose show all files in explorer, and there is indeed one .mkv movie on the drive. see below:
![Screenshot_20200226-121503.jpg](https://i.loli.net/2020/02/26/Jh9b4GLjgy6sI8M.jpg)
![Screenshot_20200226-121510.jpg](https://i.loli.net/2020/02/26/lVS5oIeWibNrPDu.jpg)

I'm not a programmer, from the log I can see Nova tries to access "/mnt/media_rw/60BB-776E", I checked several file explorer app, seems they all recognize the drive in different ways:
Mixplorer (two addresses, `/media_rw` and `storage` respectively, both accessible, can delete/write etc. BUT CANNOT trigger playback within app by NOVA, strange because this used to be worked on my TV. What's more VLC can be triggered within this app):
![Screenshot_20200226-122311.jpg](https://i.loli.net/2020/02/26/igO2d51qZPylIDU.jpg)

RS explorer: (needs permission before accessing, which is not showed when using NOVA directly, CAN trigger playback within app by NOVA)
![Screenshot_20200226-135134.jpg](https://i.loli.net/2020/02/26/y6AK1umwqa5sXvp.jpg)
![Screenshot_20200226-135145.jpg](https://i.loli.net/2020/02/26/xE8hszgFMaKDXl9.jpg)

CX file explorer:(seems the address has been "made up" and definitely not real location, CAN trigger playback within app by NOVA)
![Screenshot_20200226-135056.jpg](https://i.loli.net/2020/02/26/3cbrAfJsmtVOGvu.jpg)

No compare with Archos this time as it FC everytime I open it, seems not compatible on Android Pie anymore.
Also tried official VLC on Google Play, the file can be played normally. VLC recognize the drive as "OTG", maybe you can dig more coding info from VLC?
![Screenshot_20200226-133205.jpg](https://i.loli.net/2020/02/26/pJS4QEvAI21jrbU.jpg)

**Comment by courville** [comment]
Thus on the phone the drive is correctly detected:  this is another issue that the one reported on the tv.
I have a fix to share for the folder not accessible (fixing also some file/folder picker issues) that I will share.

**Comment by courville** [comment]
Could you please test this apk https://drive.google.com/open?id=1UjpLQQd9nxSd3EsU8AL4cQzC55aRxUPx (not solving the detection of drive on your toshiba tv)

**Comment by Neverends4** [comment]
Tried the app in the link ,results:

1. 
**My Android TV**: as you described, still no USB detection, but the drive can be accessed using "Legancy interface"-"Cloud" method and movie can be played when file manager passing url to Nova.
As I mentioned, VLC Android can detect and play this USB drive on TV without issue.

2. 
**My phone:** under phone UI the drive can be detected and showed correctly after giving USB drive permission to Nova (newly prompt, I think you updated the code right?):
![Screenshot_20200226-230349.jpg](https://i.loli.net/2020/02/27/CIZySu7EvR6NQTk.jpg)

But under TV UI you press the external drive icon still shows nothing:
![Screenshot_20200226-230323.jpg](https://i.loli.net/2020/02/27/wCaNpfXvHl7DVEg.jpg)

[Log here](https://we.tl/t-eKnUKpXcZz)

**Comment by courville** [comment]
The issue for android tv is that storage access framework (SAF) relies on folder/file picker which is absent from the framework on many devices. This is a huge lack that has been reported to google that needs fixing. Workaround on nova could be to rely on another external lib that does this job.

**Comment by Neverends4** [comment]
OK, so I get it the Android TV detection bug can be fixed? because it's knida annoying when using USB drive with Nova as it is now.

**Comment by bkarlson** [comment]
Hello,
I have the same problem. Using Nova on an XGIMI projector, and USB drives when connected appear under /mnt/usb/ folder. Nova should have access to this folder (just like any other app), but in Network and Files I can see only Network | Internal | Playlists sections ,no USB drives. Internal leads to /storage/emulated/0 and I cannot go any level higher than that. Any fix coming?

**Comment by venkatx5** [comment]
Is there any fix? I updated to latest version as of Sep 24, 2024. But still the issue persists.

@courville Is there any command I can run via ADB to get special access to Nova Player to read the new files on external drive?

**Comment by courville** [comment]
@venkatx5 please try to see if your external usb drive files are indexed by Android framework using:
```
adb shell content query --uri "content://media/external/video/media" --projection _id:_data
```
If files are not visible, you will need to find a way to index external drives (on some devices, it exists in android settings themselves on storage).

EDIT: cf. "Visibility of external USB drive on Nvidia Shield" section in https://home.courville.org/nova_video_player-faq/index.html
